tensorflow/python/data/experimental/ops/optimization_options.py

```diff
@@ -126,6 +126,14 @@ class OptimizationOptions(options.OptionsBase):
       "Whether to fuse filter dataset that predicts random_uniform < rate into "
       "a sampling dataset. If None, defaults to False.")
 
+  hoist_discard = options.create_option(
+      name="hoist_discard",
+      ty=bool,
+      docstring=
+      "Whether to hoist ops that will discard data (such as skip, take, shard) "
+      "out of unary cardinality preserved transformations. "
```

Comment:
Add a basic example, e.g. "dataset.map(...).take(3) gets optimized to dataset.take(3).map(...)".